### PR TITLE
[Kernel] Avoid input copies for cropped Snake activations

### PR DESCRIPTION
**Pull request by BANANASJIM**
## Purpose

`AliasFreeActivation1d` passes the cropped output of `UpSample1d` to Snake/SnakeBeta. This view has a contiguous time axis but gaps between channels; the existing Triton wrapper copies the entire input with `.contiguous()` before each activation.

Read the view directly by separating input and output batch/channel strides. Keep the output contiguous and retain materialization when the time-axis stride is not one. Kernel arithmetic, parameter caches, and eager fallback behavior are unchanged.

The runtime change is limited to `common/snake_activation.py`. Add layout/CUDA Graph tests, an operator/component microbenchmark, and a pretrained BigVGAN comparison benchmark.

## Test Plan

Requires an installed vLLM-Omni development/test environment and a CUDA GPU. From the checkout root:

```bash
export PYTHONPATH=.
python -m pytest tests/model_executor/models/common/test_snake_activation.py \
  tests/model_executor/models/qwen3_tts/test_cuda_graph_decoder.py::test_snakebeta_triton_vs_eager \
  -m 'core_model and cuda' --run-level core_model

compute-sanitizer --tool memcheck --error-exitcode 1 \
  python -m pytest tests/model_executor/models/common/test_snake_activation.py -q \
  -k '(cropped and (shape1 or shape2) and SnakeBeta) or alias_free_activation_cuda_graph'

python benchmarks/kernels/benchmark_snake_activation.py --dtype float32 --include-compiled
python benchmarks/kernels/benchmark_snake_activation.py --dtype float16 --include-compiled
python benchmarks/kernels/benchmark_snake_activation.py --dtype bfloat16 --include-compiled

# Requires network/model download in addition to the above dependencies.
git show f1a6e7ce:vllm_omni/model_executor/models/common/snake_activation.py > /tmp/snake_before.py
python benchmarks/kernels/benchmark_snake_decoder.py \
  --baseline-source /tmp/snake_before.py \
  --revision 633ff708ed5b74903e86ff1298cf4a98e921c513
```

The new tests carry `core_model` and `cuda` markers and are included in the existing CUDA ready pipeline's `Engine&Model Executor Test` sweep. No CI routing changes.

**vLLM Version:** 0.29.0 on A100 SXM4 40GB (Vast, `vllm/vllm-openai:v0.29.0`), PyTorch 2.13.0+cu130, Triton 3.7.1. Additional local RTX 3080 Ti 12GB validation used vLLM 0.28.0 with the same PyTorch/Triton versions. The editable development version emits a vLLM/Omni version-alignment warning; these results qualify the tested operators/decoder, not the full serving stack.

**vLLM-Omni Commit:** base `f1a6e7ce`; candidate `e035878f`. GPU tests and microbenchmarks ran at `65ac6d54`; later changes add the decoder benchmark and a one-line assertion rejecting Triton fallback. The A100 decoder script was byte-compared against `423f3dce`; the final assertion was then exercised successfully in a pinned-weight, 16-frame smoke run on RTX 3080 Ti (finite, bit-exact output). Runtime code is unchanged since `65ac6d54`.

### Correctness and checks

- **94 passed on both GPUs:** 76 new tests plus 18 existing SnakeBeta eager-reference cases. Covers Snake/SnakeBeta, FP32/FP16/BF16, contiguous/cropped/transposed/stepped layouts, unchanged input storage, contiguous output, and changed-input CUDA Graph replay.
- Copied-input versus view-input Triton results are bit-exact. Against the original base, the new suite gives **75 passed, 1 failed**: the regression detects `aten::contiguous` on the real upsampled view.
- The new randomized-alpha test uses `2e-3` FP16 eager-reference tolerances. An initial `1e-3` threshold failed identically on baseline and candidate due to existing fused/eager intermediate rounding. The exact copied/view comparison and existing tests are unchanged.
- Focused Compute Sanitizer memcheck on RTX 3080 Ti: **9 passed, 67 deselected, 0 errors**, including multi-row crops, the masked 4097-element length, all three dtypes, and CUDA Graph replay. A broader initial sanitizer run was interrupted for runtime; it is not counted as a pass.
- All applicable changed-file pre-commit hooks, including mypy, passed; `git diff --check` passed. The added benchmark guard also passed local pre-commit and the pretrained smoke run.

### Operator and alias-free component

A100, batch 1, 64 channels, seed 42. Five prewarm calls, then three alternating-order rounds; each uses CUDA Graph replay with `rep=100 ms`, excluding compilation/capture. Times are medians across rounds in microseconds. Input length is before 2× upsampling.

The copy baseline explicitly materializes the input before the same Triton kernel; it reproduces the previous copy path, rather than timing the original commit's kernel. The alias-free scope includes real upsample → activation → downsample operations.

| Dtype | Input length | Activation copy → view (µs) | Alias-free copy → view (µs) |
| --- | ---: | ---: | ---: |
| float32 | 128 | 3.734 → 1.766 | 20.680 → 18.697 |
| float32 | 512 | 4.306 → 2.123 | 22.662 → 20.281 |
| float32 | 2048 | 9.175 → 6.376 | 37.633 → 35.057 |
| float32 | 8192 | 13.002 → 7.994 | 77.901 → 72.967 |
| float16 | 128 | 4.094 → 1.675 | 22.383 → 19.624 |
| float16 | 512 | 4.966 → 2.149 | 24.068 → 20.755 |
| float16 | 2048 | 8.796 → 5.731 | 38.758 → 35.510 |
| float16 | 8192 | 11.601 → 6.666 | 79.094 → 73.716 |
| bfloat16 | 128 | 4.089 → 1.673 | 22.457 → 19.730 |
| bfloat16 | 512 | 5.052 → 2.140 | 24.213 → 21.029 |
| bfloat16 | 2048 | 8.809 → 5.946 | 39.031 → 35.918 |
| bfloat16 | 8192 | 11.570 → 6.682 | 80.070 → 74.763 |

Across this sweep: **1.44–2.44×** activation speedup and **1.07–1.16×** alias-free component speedup. Local RTX 3080 Ti FP32 measurements also improved across the four lengths (activation 1.50–2.18×, component 1.055–1.102×).

The isolated operator saves one activation-sized temporary; peak allocation for the complete alias-free module was unchanged. Optional `torch.compile` native baselines passed numerical checks and were faster than this explicit Triton implementation for some shapes, including length 2048. This PR does not claim superiority over every compiled alternative.

<details>
<summary>Across-round mean ± sample standard deviation (n=3)</summary>

| Dtype | Scope | Input length | Copy (µs) | View (µs) |
| --- | --- | ---: | ---: | ---: |
| float32 | activation | 128 | 4.008 ± 0.474 | 1.862 ± 0.252 |
| float32 | alias_free | 128 | 20.680 ± 0.002 | 18.697 ± 0.002 |
| float32 | activation | 512 | 4.298 ± 0.014 | 2.123 ± 0.000 |
| float32 | alias_free | 512 | 22.661 ± 0.001 | 20.281 ± 0.002 |
| float32 | activation | 2048 | 9.173 ± 0.005 | 6.375 ± 0.001 |
| float32 | alias_free | 2048 | 37.695 ± 0.129 | 35.066 ± 0.015 |
| float32 | activation | 8192 | 13.011 ± 0.031 | 7.992 ± 0.008 |
| float32 | alias_free | 8192 | 77.890 ± 0.036 | 72.993 ± 0.047 |
| float16 | activation | 128 | 4.488 ± 0.684 | 1.807 ± 0.229 |
| float16 | alias_free | 128 | 22.384 ± 0.003 | 19.624 ± 0.001 |
| float16 | activation | 512 | 4.959 ± 0.130 | 2.149 ± 0.000 |
| float16 | alias_free | 512 | 24.068 ± 0.001 | 20.756 ± 0.001 |
| float16 | activation | 2048 | 8.796 ± 0.000 | 5.731 ± 0.000 |
| float16 | alias_free | 2048 | 38.758 ± 0.002 | 35.510 ± 0.001 |
| float16 | activation | 8192 | 11.629 ± 0.049 | 6.665 ± 0.002 |
| float16 | alias_free | 8192 | 79.045 ± 0.086 | 73.682 ± 0.058 |
| bfloat16 | activation | 128 | 4.483 ± 0.683 | 1.832 ± 0.275 |
| bfloat16 | alias_free | 128 | 22.456 ± 0.002 | 19.729 ± 0.002 |
| bfloat16 | activation | 512 | 5.016 ± 0.189 | 2.141 ± 0.000 |
| bfloat16 | alias_free | 512 | 24.213 ± 0.000 | 21.030 ± 0.002 |
| bfloat16 | activation | 2048 | 8.809 ± 0.000 | 5.946 ± 0.002 |
| bfloat16 | alias_free | 2048 | 39.035 ± 0.011 | 35.921 ± 0.007 |
| bfloat16 | activation | 8192 | 11.589 ± 0.034 | 6.681 ± 0.002 |
| bfloat16 | alias_free | 8192 | 80.039 ± 0.055 | 74.784 ± 0.038 |

</details>

### Pretrained decoder

Run the real IndexTTS2 BigVGAN implementation with `nvidia/bigvgan_v2_22khz_80band_256x` at the pinned revision above, FP32, batch 1, synthetic log-mel inputs (seed 42). Both models use identical pretrained weights; replace only the 109 Snake activations in the baseline copy with the original source from `f1a6e7ce`.

The benchmark checks regenerated fixed filters against checkpoint filters within `1e-7`, removes only those non-persistent filter entries, then strictly loads all remaining weights. Ten warmups precede three alternating-order CUDA Graph timing rounds with `rep=200 ms`.

| Mel frames | Output samples | Original (ms) | Candidate (ms) | Latency reduction | Max absolute difference |
| ---: | ---: | ---: | ---: | ---: | ---: |
| 16 | 4096 | 7.591 | 7.319 | 3.59% | 0 |
| 64 | 16384 | 13.290 | 12.839 | 3.39% | 0 |
| 256 | 65536 | 31.007 | 29.843 | 3.75% | 0 |

All three outputs were finite and bit-exact. These are complete vocoder graph-replay measurements with synthetic mel inputs, not full text-to-speech serving or perceptual audio evaluation.

Across-round decoder mean ± sample standard deviation (n=3):

| Mel frames | Original (ms) | Candidate (ms) |
| ---: | ---: | ---: |
| 16 | 7.784 ± 0.336 | 7.319 ± 0.002 |
| 64 | 13.290 ± 0.005 | 12.839 ± 0.001 |
| 256 | 31.013 ± 0.016 | 29.839 ± 0.007 |

The original decoder at 16 frames had a slower first measured round (8.172 ms versus 7.591/7.590 ms); all samples are retained below. Three rounds provide preliminary evidence, not a statistical guarantee.

<details>
<summary>Decoder per-round milliseconds</summary>

```json
[
  {
    "mel_frames": 16,
    "original": [
      8.171929550170898,
      7.591116714477539,
      7.590195083618164
    ],
    "candidate": [
      7.318783760070801,
      7.320661544799805,
      7.317503929138184
    ]
  },
  {
    "mel_frames": 64,
    "original": [
      13.289983367919922,
      13.295308685302734,
      13.285785675048828
    ],
    "candidate": [
      12.8405335744222,
      12.83758087158203,
      12.839082717895508
    ]
  },
  {
    "mel_frames": 256,
    "original": [
      31.03057861328125,
      31.00661773681641,
      31.00098571777344
    ],
    "candidate": [
      29.83055419921875,
      29.843455505371097,
      29.843457031249997
    ]
  }
]
```
</details>

### Independent automated review

A separate reviewer inspected final head `e035878f` for correctness, minimality, project conventions, CI selection, and evidence provenance. No unresolved findings. One optional suggestion resulted in the single post-timing Triton guard, verified by the pinned-weight smoke run. This is automated review evidence; maintainer approval remains outstanding.

### Remaining qualification

Full TTS serving latency/concurrency, real mel workloads, perceptual audio quality, and ROCm have not been tested. Human/maintainer review remains outstanding; no end-to-end speech-generation speedup is claimed.

AI assistance: Codex implemented the patch, tests and benchmarks, ran validation, and prepared this description.

### CODE REVIEW

**Review by Dong1017 — COMMENTED** [review]
Reviewed head d062ba24a419 against main ae3880f0e39d. I found no blocking code issue. The kernel reads cropped input with its batch/channel strides and writes through separate contiguous-output strides; the input pointer includes the crop's storage offset. Time-stride != 1 materialization, tail masking, per-channel arithmetic, and eager fallback are unchanged. The new layout, real UpSample1d crop, and CUDA Graph tests target these contracts.

The author reports 94 GPU passes and focused memcheck at 65ac6d54. `snake_activation.py` has the same Git blob at that tested commit and at d062ba24; I have not independently run GPU tests on the merged head. Visible checks pass, but do not show CUDA tests. The reported operator, alias-free component, and pretrained BigVGAN gains measure CUDA Graph replay (synthetic mel for BigVGAN), not full TTS serving. No code-level blocker to continued maintainer review was found. Please add the short author self-review requested by the review bot.
